R/particle_filter_state.R

```diff
@@ -0,0 +1,220 @@
+##' @title Particle filter state
+##'
+##' @description Particle filter internal state. This object is not
+##'   ordinarily constructed directly by users, but via the
+##'   `$run_begin` method to [mcstate::particle_filter]. It provides
+##'   an advanced interface to the particle filter that allows
+##'   partially running the particle filter over part of the time
+##'   trajectory.
+##'
+##' This state object has a number of public fields that you can read
+##'   but must not write (they are not read-only so you *could* write
+##'   them, but don't).
+particle_filter_state <- R6::R6Class(
+  "particle_filter_state",
+  cloneable = FALSE,
+
+  private = list(
+    generator = NULL,
+    pars = NULL,
+    data = NULL,
+    data_split = NULL,
+    steps = NULL,
+    n_particles = NULL,
+    n_threads = NULL,
+    initial = NULL,
+    index = NULL,
+    compare = NULL,
+    save_restart_step = NULL,
+    save_restart = NULL,
+    current_step_index = 0L
+  ),
+
+  public = list(
+    ##' @field model The dust model generator being simulated (cannot be
+    ##'   re-bound)
+    model = NULL,
+
+    ##' @field complete Logical, indicating if the particle filter has
+    ##'   reached the end of the data.
+    complete = FALSE,
+
+    ##' @field history The particle history, if created with
+    ##'   `save_history = TRUE`. This is an internal format subject to
+    ##    change.
+    history = NULL,
+
+    ##' @field restart_state Full model state at a series of points in
+    ##'   time, if the model was created with non-`NULL` `save_restart`.
+    ##'   This is a 3d array as described in [mcstate::particle_filter]
+    restart_state = NULL,
+
+    ##' @field log_likelihood The log-likelihood so far. This starts at
+    ##'   0 when initialised and accumulates value for each step taken.
+    log_likelihood = NULL,
+
+    ##' @description Initialise the particle filter state. Ordinarily
+    ##' this should not be called by users, and so arguments are not
+    ##' documented.
+    initialize = function(pars, generator, model, data, data_split, steps,
+                          n_particles, n_threads, initial, index, compare,
+                          seed, save_history, save_restart) {
+      ## NOTE: this will generate a warning when updating docs but
+      ## that's ok; see https://github.com/r-lib/roxygen2/issues/1067
+      if (is.null(model)) {
+        model <- generator$new(pars = pars, step = steps[[1L]],
+                               n_particles = n_particles, n_threads = n_threads,
+                               seed = seed)
```

Comment:
Is it now possible for us to use a single dust model for all of the particle filters here, by using the index of parameter membership? Or is that too complex to begin with, and we should do that in a future PR?

Reply:
We should definitely explore that once we get things working. Will be interesting to see when that's faster than running them sequentially or lower-core numbers of particle filters in parallel....